### PR TITLE
Translation improvements: CLDR Plurals + Translatables

### LINKED ISSUE #23797: Incorrect handling of plural forms for translation
### Description

Hi!

The current handling of strings with multiple plural forms by mapping them to key-value pairs in ini file is incorrect. It allows using only two plural forms like in English (singular for one only). There's also no problems that have only one plural form. But there are many languages with three and more plural forms (you may find a review [here][1] for example). Correct translation to such languages is currently impossible.

Please consider changing the translation framework to handle pluralization properly.

[1]: https://www.gnu.org/software/gettext/manual/html_node/Plural-forms.html

### Gitea Version

1.19.0

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

n/a

### Database

None

### PR DESCRIPTION
**Pull request by zeripath**
This PR will provide the framework for a few translation improvements.

## `TranslatableStringer` and `TranslatableFormatted`

The first thing is the ability to pass in translatable elements - this would allow us (for example) to create an error:

```go
type TranslatableError struct {
Key string
Args []interface{}
}

func (err *TranslatableError) Error() string {
   return i18n.Tr("en-US", err.Key, err.Args...)
}

func (err *TranslatableError) TranslatedString(l i18n.Locale) {
   return l.Tr(err.Key, err.Args...)
}
```

## `TrPlural` and `TrOrdinal`

The second part is the creation of `TrPlural` and `TrOrdinal` which use the CLDR data to correctly determine plurals for the locales. This is used to replace `TrN`. 

Plural and ordinal keys can be written as go templates if the key has the suffix `_plural` or `_ordinal`:

```ini
example_plural=common text {{if .One}}Singular{{else}}Other{{end}}
```

These are interpreted at the time of locale reading and split into the appropriate forms for the language. Alternatively the individual plural forms can be provided separately as `example_plural_one` etc. 

As the templates are parsed once at locale load time the magic suffix is required - I don't think this is too clunky. It would be possible to make it so we parse the template at time of TrPlural call but that would require the locale storing a text.Template or other locking mechanism, and I think the magic suffix isn't too inconvenient.

(`TrN` has been kept in place to help handle merges and old data.) 

Fixes #23797

### CODE REVIEW

**Comment by zeripath** [comment]
Conflicts resolved.

**Comment by zeripath** [comment]
Now... This PR as it stands relies on adding lots of additional per language keys which I think aren't handled well by crowdin.

I think there we may need to instead use a kind of template.


Instead of:

```
issue.action.push_plural_one = <b>@%[1]s</b> pushed %[3]d commit to %[2]s
issue.action.push_plural = <b>@%[1]s</b> pushed %[3]d commits to %[2]s
```

How about:

```
issue.action.push_plural = <b>@%[1]s</b> pushed %[3]d commit{{if not .One}}s{{end}} to %[2]s
```

Then we can use the go (text) template package to do the parsing.

Another option would be use a handlebarslike syntax:

```
issue.action.push_plural = <b>@%[1]s</b> pushed %[3]d {{One}}commit{{/One}}{{Other}}commits{{/Other}} to %[2]s
```

**Comment by lafriks** [comment]
How that would work with crowdin? When English have two forms but other languages could have 3 or 4?

**Comment by zeripath** [comment]
```ini
example_plural = {{if .Zero}}zero{{else if .One}}one{{else}}other{{end}}form
```

Or add each form individually as example_plural_zero etc. (But I don't think crowdin will let you do that at present.)

**Comment by zeripath** [comment]
@lafriks I've updated the description.

File: docs/content/doc/features/localization.en-us.md [hunk 1]
```diff
@@ -31,3 +31,70 @@ After a translation has been accepted, it will be reflected in the main reposito
...
+
+If the provided key does not have the `_plural` suffix or its format template is invalid, the specific per plural forms must be provided explicitly. The plural key fallback hierarchy is as follows:
+
+1. `${key}_${form}` in the locale.
+2. `${key}_other` in the locale.
+3. `${key}` in the locale.
+4. `${key}_${form}` in the default locale.
+5. `${key}_other` in the default locale.
+6. `${key}` in the default locale.
+7. Use the `${key}_${form}` directly as the format.
```
Comment (wxiaoguang) [inline docs/content/doc/features/localization.en-us.md]: Does it work with current Gitea's translation system?

For example, the keys are only for English, other languages do not have the change to have other keys?

Comment (wxiaoguang) [reply docs/content/doc/features/localization.en-us.md]: And some more thoughts: as a translator, I could only translate what I have seen. Even if I know Gitea's code, when I do translation, I do not want to add any new keys (personally), it's too difficult for me ... I doubt whether there are translators could benefit from this mechanism.

**Comment by wxiaoguang** [comment]
Generally L-G-T-M.

Could it be better if the `plurals` and the `common/dtd` are put in a new public go package? They seem quite complex and independent. Making them a new package would keep Gitea repo more focused. (and maybe other projects could re-use them)

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1009,6 +1007,7 @@ broken_message = The Git data underlying this repository cannot be read. Contact
 code = Code
 code.desc = Access source code, files, commits and branches.
 branch = Branch
+branch_plural = Branch{{if not .One}}es{{end}}
```
Comment (wxiaoguang) [inline options/locale/locale_en-US.ini]: It seems that `{{if .One}}Branch{{else}}Branches{{end}}` could be more friendly to translators.

Comment (zeripath) [reply options/locale/locale_en-US.ini]:
```suggestion
branch_plural = {{if .One}}Branch{{else}}Branches{{end}}
```

File: docs/content/doc/features/localization.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/features/localization.en-us.md]: I would not expect that we would want to provide ANY keys like this. The point is that the `_plural` and `_ordinal` suffix causes the auto-creation of these for you and therefore this should be definitely preferred.

Explaining the hierarchy is helpful though for people who want to write their own locale files.

**Comment by zeripath** [comment]
> Generally L-G-T-M.
> 
> Could it be better if the `plurals` and the `common/dtd` are put in a new public go package? They seem quite complex and independent. Making them a new package would keep Gitea repo more focused. (and maybe other projects could re-use them)

I think we should try to figure out what works before export things. for example we need to include date and time formatting at some point.

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (wxiaoguang) [reply options/locale/locale_en-US.ini]: I meant there are many `word{{if not .One}}s{{end}}` in code.

Do we prefer to keep using `word{{if not .One}}s{{end}}` or using `{{if .One}}word{{else}}words{{end}}`? Personally I would prefer to make it easier for translators.

Either format is preferred, IMO it's better to document it.

Comment (zeripath) [reply options/locale/locale_en-US.ini]: done.

File: docs/content/doc/features/localization.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/features/localization.en-us.md]: I've adjusted the wording here to make it clear that this really is TECHNICAL INFORMATION

**Comment by wxiaoguang** [comment]
We need ICU message system #23863, this homemade syntax is impossible to maintain.

**Pull request closed without merging**